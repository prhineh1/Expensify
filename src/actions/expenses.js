import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD_EXPENSE

export const addExpense = (expense = {}) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const addExpenseComplete = (expense = {}) => ({
    type: 'ADD_EXPENSE_COMPLETE',
    expense
});

//REMOVE_EXPENSE

export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//EDIT_EXPENSE

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES

export const setExpenses = () => ({ type: 'SET_EXPENSES' });

export const setExpensesComplete = (expenses) => ({
    type: 'SET_EXPENSES_COMPLETE',
    expenses
});