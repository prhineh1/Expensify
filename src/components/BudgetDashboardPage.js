import React from 'react';
import BudgetsSummary from './BudgetsSummary';
import BudgetList from './BudgetList';

const BudgetDashboardPage = () => (
  <div>
      <BudgetsSummary />
      <BudgetList />
  </div>
);

export default BudgetDashboardPage;