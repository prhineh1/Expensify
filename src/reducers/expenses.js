//Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE_COMPLETE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE_COMPLETE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE_COMPLETE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES_COMPLETE':
            return action.expenses;
        default:
            return state;
    }
};
