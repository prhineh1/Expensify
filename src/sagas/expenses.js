import ReactDOM from 'react-dom';
import { takeEvery } from 'redux-saga';
import { put, all, call } from 'redux-saga/effects'
import * as Api from '../firebase/firebase';

//WATCHERS

export function* watchAddExpenseSaga() {
    yield takeEvery('ADD_EXPENSE', addExpenseAsync);
};

export function* watchSetExpensesSaga() {
    yield takeEvery('SET_EXPENSES', setExpensesAsync);
};

export function* watchRemoveExpenseSaga() {
    yield takeEvery('REMOVE_EXPENSE', removeExpenseAsync);
};

export function* watchEditExpenseSaga() {
    yield takeEvery('EDIT_EXPENSE', editExpenseAsync);
}

//WORKERS

export function* addExpenseAsync({ expense, uid }) {
    try {
        const response = yield call(Api.create, expense, uid);  
        yield put({ 
            type: 'ADD_EXPENSE_COMPLETE', 
            expense: { id: response.key, ...expense}
        });      
    }
    catch(error) {
        return;
    }
};

export function* setExpensesAsync(uid) {
    try {
        const expenses = yield call(Api.get, uid);  
        yield put({
            type: 'SET_EXPENSES_COMPLETE',
            expenses
        });      
    }
    catch(error) {
        return;
    }
};

export function* removeExpenseAsync({ id, uid }) {
    try {
        yield call(Api.remove, id, uid);  
        yield put({
            type: 'REMOVE_EXPENSE_COMPLETE',
            id
        });      
    }
    catch(error) {
        return;
    }
};

export function* editExpenseAsync({ updates, id, uid }) {
    try {
        yield call(Api.edit, updates, id, uid); 
        yield put({
            type: 'EDIT_EXPENSE_COMPLETE',
            id,
            updates
        });       
    }
    catch(error) {
        return;
    }
};