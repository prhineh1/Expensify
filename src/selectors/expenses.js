import moment from 'moment';
import partition from 'js-partition';

//Get visible expenses

export default (expenses, {text, sortBy, startDate, endDate, pagination, expensesPartitionIndex}) => {
    let partitionedExpenses = expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return b.createdAt - a.createdAt;
        } else if (sortBy === 'amount') {
            return b.amount - a.amount;
        }
    });
    
    if (partitionedExpenses.length % pagination === 0 && partitionedExpenses.length > 0) {
        partitionedExpenses = partition(pagination, pagination, partitionedExpenses);
    }
    else {
        partitionedExpenses = partition(pagination, pagination, [], partitionedExpenses);        
    }

    //first return value is a jagged array with all the expenses
    //second return value is the current array that will be displayed
    return [partitionedExpenses, partitionedExpenses[expensesPartitionIndex]];
};