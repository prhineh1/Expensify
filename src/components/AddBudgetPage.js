import React from 'react';
import { connect } from 'react-redux';
import BudgetForm from './BudgetForm';
import { addBudget } from '../actions/budgets';

export class AddBudgetPage extends React.Component {
  onSubmit = (Budget) => {
    this.props.addBudget(Budget, this.props.uid);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
          <div className='page-header'>
              <div className='content-container'>
                  <h1 className='page-header__title'>Add Budget</h1>
                  <p className='page-header__subtitle'>Note: Budgeted expenses are tracked by month.</p>
              </div>
          </div>
          <div className='content-container'>
              <BudgetForm
              onSubmit={this.onSubmit}
              />
            </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({ uid: state.auth.uid });

export default connect(mapStateToProps, { addBudget })(AddBudgetPage);