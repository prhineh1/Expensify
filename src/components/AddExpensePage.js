import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense, this.props.uid)
        this.props.history.push('/');
    };
    render() {
        return (
        <div>
            <div className='page-header'>
                <div className='content-container'>
                    <h1 className='page-header__title'>Add Expense</h1>
                </div>
            </div>
            <div className='content-container'>
                <ExpenseForm
                onSubmit={this.onSubmit}
                budgets={this.props.budgets}
                />
             </div>
        </div>
        );
    }
}

const mapStateToProps = (state) => ({ 
    uid: state.auth.uid,
    budgets: state.budgets
});

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense, uid) => dispatch(addExpense(expense, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);