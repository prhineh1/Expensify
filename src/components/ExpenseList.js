import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import { setPartitionIndex } from '../actions/filters';

export const ExpenseList = (props) => (
    <div className='content-container'>
        <div className='list-header'>
            <div className='show-for-mobile'>Expenses</div>
            <div className='show-for-desktop'>Expense</div>
            <div className='show-for-desktop'>Budget</div>
            <div className='show-for-desktop'>Amount</div>
        </div>
        <div className='list-body'>
            {
                props.expenses.length === 0 ? (
                    <div className='list-item list-item--message'>
                        <span>No expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense, index) => 
                    <ExpenseListItem key={expense.id} expense={expense} budgets={props.budgets} />
                    )
                )
            }
        </div>
        {
            !!props.partitions && (
                <div className='list-nav'>
                    {
                        props.partitions[props.next] && (
                            <button className='button button--nav' onClick={() => props.setPartitionIndex(props.next)}>Next</button>
                        )
                    }
                    { 
                        props.partitions[props.prev] && (
                            <button className='button button--nav' onClick={() => props.setPartitionIndex(props.prev)}>Previous</button>
                        )
                    }
                </div>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    const [partitionedExpenses, selectedExpenses] = selectExpenses(state.expenses, state.filters);
    return {
        expenses: selectedExpenses,
        partitions: partitionedExpenses,
        next: state.filters.expensesPartitionIndex + 1,
        prev: state.filters.expensesPartitionIndex - 1,
        budgets: state.budgets
    };
};

export default connect(mapStateToProps, { setPartitionIndex })(ExpenseList);

