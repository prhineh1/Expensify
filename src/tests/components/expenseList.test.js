import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';


test('should render ExpenseList with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} next={1} prev={-1} partitions={[expenses]} />)
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} next={1} prev={-1} partitions={[]} />)
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with partitioned expenses', () => {
    const partitionedExpenses = [ [expenses[0], expenses[1]], [expenses[2]] ];
    const wrapper = shallow(<ExpenseList expenses={expenses} next={1} prev={0} partitions={partitionedExpenses} />);
    expect(wrapper).toMatchSnapshot();  
});

test('should handle next button click event', () => {
    const partitionedExpenses = [ [expenses[0], expenses[1]], [expenses[2]] ]; 
    const setPartitionIndex = jest.fn();
    const wrapper = shallow(<ExpenseList setPartitionIndex={setPartitionIndex} 
                                        expenses={expenses} 
                                        next={1} prev={0} 
                                        partitions={partitionedExpenses} />);
    wrapper.find('button').at(0).simulate('click');   
    expect(setPartitionIndex).toHaveBeenCalledWith(1); 
});

test('should handle previous button click event', () => {
    const partitionedExpenses = [ [expenses[0], expenses[1]], [expenses[2]] ]; 
    const setPartitionIndex = jest.fn();
    const wrapper = shallow(<ExpenseList setPartitionIndex={setPartitionIndex} 
                                        expenses={expenses} 
                                        next={1} prev={0} 
                                        partitions={partitionedExpenses} />);
    wrapper.find('button').at(1).simulate('click');   
    expect(setPartitionIndex).toHaveBeenCalledWith(0); 
});