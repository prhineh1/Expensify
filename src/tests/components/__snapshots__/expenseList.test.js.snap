// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Budget
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No expenses
      </span>
    </div>
  </div>
  <div
    className="list-nav"
  />
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Budget
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      expense={
        Object {
          "amount": 195,
          "budgetId": "asf0923",
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      key="1"
    />
    <ExpenseListItem
      expense={
        Object {
          "amount": 10095,
          "budgetId": "a;sdlfj2089",
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      key="2"
    />
    <ExpenseListItem
      expense={
        Object {
          "amount": 4500,
          "budgetId": "tuq3948tu",
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": "3",
          "note": "",
        }
      }
      key="3"
    />
  </div>
  <div
    className="list-nav"
  />
</div>
`;

exports[`should render ExpenseList with partitioned expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Budget
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      expense={
        Object {
          "amount": 195,
          "budgetId": "asf0923",
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      key="1"
    />
    <ExpenseListItem
      expense={
        Object {
          "amount": 10095,
          "budgetId": "a;sdlfj2089",
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      key="2"
    />
    <ExpenseListItem
      expense={
        Object {
          "amount": 4500,
          "budgetId": "tuq3948tu",
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": "3",
          "note": "",
        }
      }
      key="3"
    />
  </div>
  <div
    className="list-nav"
  >
    <button
      className="button button--nav"
      onClick={[Function]}
    >
      Next
    </button>
    <button
      className="button button--nav"
      onClick={[Function]}
    >
      Previous
    </button>
  </div>
</div>
`;
